PULP Event Load Extension Instructions + GNU Tests Version 1.0.0

NandniJamnadas opened the pull request.

NandniJamnadas committed:
PULP: Added Event Load Instructions + Tests Version 1.0.0 extension in assembler

       The event load instruction cv.elw is only supported if the
       PULP_CLUSTER parameter is set to 1. The event load performs a
       load word and can cause the CV32E41P to enter a sleep state

        bfd/
            * elfxx-riscv.c: (riscv_multi_subset_supports): Added
              INSN_CLASS_XCOREVELW.
            * elfxx-riscv.c: (riscv_multi_subset_supports_ext): Likewise.

        gas/
            * testsuite/gas/riscv/cv-elw-pass.d: Added supported instruction
              test.
            * testsuite/gas/riscv/cv-elw-pass.s: Likewise.
            * testsuite/gas/riscv/cv-elw-fail.d: Added unsupported
              instruction test.
            * testsuite/gas/riscv/cv-elw-fail.s: Likewise.
            * testsuite/gas/riscv/cv-elw-pass.l: Likewise.

        include/
            * opcode/riscv-opc.h: Added corresponding MATCH and MASK
              instruction opcode macros.
            * opcode/riscv-opc.h: Added corresponding DECLARE_INSN.
            * opcode/riscv.h: (riscv_insn_class) Added
              INSN_CLASS_COREV_ELW.

        opcodes/
            * riscv-opc.c: (riscv_opcode) Added event load
            instructions.

Signed-off-by: NandniJamnadas <nandni.jamnadas@embecosm.com>

NandniJamnadas committed:
Event Load: Documentation Updated

	* gas/doc/c-riscv.texi: Updated c-riscv.texi document to
	include CORE-V Event Load extension

Signed-off-by: NandniJamnadas <nandni.jamnadas@embecosm.com>

jeremybennett reviewed (changes requested):
Overall, looks good. A couple of changes

1. You are lacking a GNU compliant commit message for the documentation commit.
2. You need some assembler tests of the validity of the extension.
    - is ELW only enabled when `-march=...Xcorevelw`
 3. You should include your regression test results as a comment on the PR.

jeremybennett commented on bfd/elfxx-riscv.c:
```diff
@@ -2622,6 +2625,8 @@ riscv_multi_subset_supports_ext (riscv_parse_subset_t *rps,
     case INSN_CLASS_COREV_BI:
       return "xcorevbi' or `xcorev";
+    case INSN_CLASS_COREV_ELW:
+      return "xcorevelw' or `xcorev";
```
TBD - what does `corev` mean when we have new versions of the tool chain. Suggestion
- `corev1p0`: Version 1 of all 5 CV32E40Pv1 extensions
- `corev2p0`: Version 2 of all 5 CV32E40Pv1 extensions and version ` of all C32E40Pv2 extensions

jeremybennett commented on gas/doc/c-riscv.texi:
```diff
@@ -50,15 +50,17 @@ is provided to enable all CORE-V extensions. To enable only a subset of the CORE
+immediate branching: @samp{Xcorevbi} (version 1 for CV32E40Pv1, version 2 for CV32E40Pv2)
+@item
+event load: @samp{Xcorevelw} (not supported for CV32E40Pv1, version 1 for CV32E40Pv2)
 @end itemize
```
Add `corev` as a catchall.

jeremybennett replied:
All fixed.

jeremybennett commented on gas/testsuite/gas/riscv/cv-elw-fail.s:
```diff
@@ -0,0 +1,8 @@
+target:
+    # Immediate Boundary Tests
+    cv.elw x5,-2049(x6)
+    cv.elw x5,2048(x6)
```
What about negative offsets?

NandniJamnadas committed:
PULP: Added Event Load MARCH GNU assembler tests

       	 * gas/testsuite/gas/riscv/cv-elw-march-fail.d: Added cv.elw
       	 instruction tests against MARCH=rv32i
	 * gas/testsuite/gas/riscv/cv-elw-march-fail.l: Likewise
	 * gas/testsuite/gas/riscv/cv-elw-march-fail.s: Likewise
	 * gas/testsuite/gas/riscv/cv-elw-march-xcorev.d: Added
	 cv.elw instruction tests against MARCH=rv32i_xcorev1p0
	 * gas/testsuite/gas/riscv/cv-elw-march-xcorev.s: Likewise

Signed-off-by: NandniJamnadas <nandni.jamnadas@embecosm.com>

NandniJamnadas commented:
1. "You are lacking a GNU compliant commit message for the documentation commit":
Updated the documentation commit message: https://github.com/openhwgroup/corev-binutils-gdb/pull/50/commits/a9405b6ac6ac60f26acbfe3cd280b4e8764ed4bb
2. "You need some assembler tests of the validity of the extension":
 Added xcorev and xcorevelw tests: https://github.com/openhwgroup/corev-binutils-gdb/pull/50/commits/6f184a9a70ec5fecde3adfec79776896fd3dcce2
3. "You should include your regression test results as a comment on the PR".
GNU Assembler Results:
```
| Category             | Baseline | Curr. Week |       Delta |
| :------------------- | -------: | ---------: | ----------: |
| Expected passes      |      785 |        788 |          +3 |
| Unexpected failures  |        - |          - |           - |
| Unexpected successes |        - |          - |           - |
| Expected failures    |       23 |         23 |           - |
| Unresolved testcases |        - |          - |           - |
| Unsupported tests    |        9 |          9 |           - |
```

jeremybennett commented:
All looks good. OK to merge.
